### PR TITLE
Release tweaks

### LINKED ISSUE #195: Bottleneck KPI
Add a KPI which analyzes number of item relations, both incoming and outgoing.

* rework scaling KPI (it is too superficial)
* differentiate between provider relations (hard) and data flow (soft)

 * red if 0 as provider for other items
 * yellow if scaled to 0 without relations
 * orange of scaled to 0 as data sink
 * unknown if no label or not a number
 * green if scaled higher than 1
 * yellow if a bottleneck (more than 1 item depend on it)

### LINKED ISSUE #354: Prepare release 0.3.4
- [x] update screenshots
- [x] check docs
- [x] changelog
- [x] gui: hide labels with masked secrets (not useful)
- [x] gui:abbreviate long label values (e.g. readme content)

### PR DESCRIPTION
**Pull request by bonndan**
addresses #354 
closes #195

### CODE REVIEW

**Comment by mfbieber** [comment]
I noticed that the text conted is cropped. Maybe we could display the content at least when hovering over it?

sidecard

**Comment by mfbieber** [comment]
Also I noticed that the icons in the graph are not always clickable (they don't have the hand symbol and nothing happens if you click on it). I am not able to reproduce it in a reliable way and it happens with different icons.

I think we might have a bug somewhere causing this 😞

**Comment by mfbieber** [comment]
In a previous PR we talked about the clickability of lines (relations). Not all show the hand symbol. For example in http://localhost:3000/#/landscape/nivio:example the blue dashed line (dashes are triangles) is not clickable, but the others are. Why is that?

**Comment by mfbieber** [comment]
> Also I noticed that the icons in the graph are not always clickable (they don't have the hand symbol and nothing happens if you click on it). I am not able to reproduce it in a reliable way and it happens with different icons.
> 
> I think we might have a bug somewhere causing this 😞

Actualy now I know how to reproduce it: 

1. reload the page (http://localhost:3000/#/landscape/nivio:example)
2. click on the relation between wordpress-redis and demo-blog
3. try to click on worpress-redis

**Comment by mfbieber** [comment]
Another thing I noticed is e.g. the information in the demo-dashboard relation. There is a space after outbound and before of the comma:
sidecard2

File: src/main/app/src/Components/Landscape/Modals/Item/Item.tsx [hunk 1]
```diff
@@ -151,24 +170,45 @@ const Item: React.FC<Props> = ({ useItem, locateItem, fullyQualifiedItemIdentifi
...
+              aria-controls='panel1a-content'
+              id='panel1a-header'
+            >
+              more
+            </AccordionSummary>
+            <AccordionDetails>
+              <div className='labels'>{item ? getLabels(item) : null}</div>
+            </AccordionDetails>
+          </Accordion>
+
```
Comment (mfbieber) [inline src/main/app/src/Components/Landscape/Modals/Item/Item.tsx:211]: This looks so much better now and really nice :)

File: src/main/java/de/bonndan/nivio/assessment/kpi/ScalingKPI.java [hunk 2]
```diff
@@ -1,26 +1,92 @@
...
+
+        if (scaleValue == 1 && (usedAsProvider > 1)) {
+            return List.of(new StatusValue(Label.scale.name(), Status.YELLOW, String.format("Unscaled, but %d items depend on it.", usedAsProvider)));
+        }
+
+        if (scaleValue > 0) {
+            return List.of(new StatusValue(Label.scale.name(), Status.GREEN));
+        }
+
+        return List.of();
```
Comment (mfbieber) [inline src/main/java/de/bonndan/nivio/assessment/kpi/ScalingKPI.java:83]: I prefer `Collections.emptyList()`, consider a possible NPE as described here: https://stackoverflow.com/questions/39400238/list-of-or-collections-emptylist

File: src/main/java/de/bonndan/nivio/util/URIHelper.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package de.bonndan.nivio.util;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Optional;
+
+public class URIHelper {
+
+    /**
+     * Returns an URI object if the string is an URI and has host and scheme.
```
Comment (mfbieber) [inline src/main/java/de/bonndan/nivio/util/URIHelper.java:10]: Actually it returns an Optional of an URI object ;)

File: src/test/java/de/bonndan/nivio/assessment/kpi/ScalingKPITest.java [hunk 4]
```diff
@@ -4,35 +4,135 @@
...
+import de.bonndan.nivio.model.RelationType;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
 import java.util.List;
+import java.util.Set;
 
 import static org.junit.jupiter.api.Assertions.*;
 
 class ScalingKPITest {
```
Comment (mfbieber) [inline src/test/java/de/bonndan/nivio/assessment/kpi/ScalingKPITest.java:17]: I did not go trough all the test cases in detail, but very nice 😻

File: src/main/java/de/bonndan/nivio/util/URIHelper.java (hunk 3 above)
Comment (bonndan) [reply src/main/java/de/bonndan/nivio/util/URIHelper.java:10]: done

File: src/main/java/de/bonndan/nivio/assessment/kpi/ScalingKPI.java (hunk 2 above)
Comment (bonndan) [reply src/main/java/de/bonndan/nivio/assessment/kpi/ScalingKPI.java:83]: Wow, good to know. Fixed.

**Comment by bonndan** [comment]
I've opened https://github.com/dedica-team/nivio/issues/361

**Comment by bonndan** [comment]
@mfbieber everything should be fixed now

File: docs/source/env_config.rst [hunk 5]
```diff
@@ -0,0 +1,22 @@
+* **DEMO** A non-empty value causes nivio to start in demo mode with prepared data.
+* **GITHUB_JWT** To connect via JWT token as a GitHub App:
+* **GITHUB_LOGIN** GitHub user name (can also be used to connect as organisation with oauth
+* **GITHUB_OAUTH** To connect to GitHub via Personal access token
+* **GITHUB_PASSWORD** GitHub password (for username/pw login)
+* **GITLAB_HOST_URL** The full URL to the GitLab API, e.g. http://your.gitlab.server.com
+* **GITLAB_PASSWORD** GitLab OAuth login password (optional)
+* **GITLAB_PERSONAL_ACCESS_TOKEN** the personal token to access the GitLab API (optional)
+* **GITLAB_USERNAME** GitLab OAuth login username (optional, if used password is also required)
+* **KUBERNETES_MASTER** K8s master url (optional), all variables from https://github.com/fabric8io/kubernetes-client#configuring-the-client can be used
```
Comment (mfbieber) [inline docs/source/env_config.rst:10]: I still think that this suggests that all k8s variables are supported right now. As long as this does not really go into the release and the k8s support will also be delivered with the next release, this is fine ;)

**Comment by mfbieber** [comment]
> In a previous PR we talked about the clickability of lines (relations). Not all show the hand symbol. For example in http://localhost:3000/#/landscape/nivio:example the blue dashed line (dashes are triangles) is not clickable, but the others are. Why is that?

Now it is clickable, but the hand symbol is not shown.

File: docs/source/env_config.rst (hunk 5 above)
Comment (mfbieber) [reply docs/source/env_config.rst:10]: We talked about this directly, k8s support is alpha, but the library we use already supports the mentioned environment variables.

**Comment by bonndan** [comment]
Is the pointer cursor maybe only show if hovering over the triangles? #360 will improve this

**Comment by mfbieber** [comment]
I think #361 is resolved with this PR.

**Comment by mfbieber** [comment]
> Is the pointer cursor maybe only show if hovering over the triangles? #360 will improve this

No, also then it is not displayed :-/

**Comment by mfbieber** [comment]
Nice that you implemented those changes after the suggestions. If #360 will address the triangle-dashed-line, we can close this!

**Pull request merged**